src/runtype.ts

```diff
@@ -99,9 +104,13 @@ export interface Runtype<A = unknown> {
  */
 export type Static<A extends Runtype> = A['_falseWitness'];
 
-export function create<A extends Runtype>(validate: (x: any) => Result<Static<A>>, A: any): A {
+export function create<A extends Runtype>(
+  validate: (x: any, visited: VisitedState, self: A) => Result<Static<A>>,
+  A: any,
+): A {
   A.check = check;
-  A.validate = validate;
+  A.innerValidate = (x: any, visited: VisitedState) => validate(x, visited, A);
+  A.validate = (x: any) => validate(x, VisitedState(), A);
```

Comment:
It doesn't seem like we need to make every single runtype responsible for making these checks, we can just do it centrally here:

```ts
  A.validate = (x: any, visited: VisitedState = VisitedState()) => {
    if (visited.has(x, A)) return { success: true, value: x };
    return validate(x, visited);
  };
```

I also don't think `innerValidate` is useful, we just need an optional extra parameter on `validate`.

Reply:
That's true. I was thinking that we couldn't do that because of Union, but it looks like while Union doesn't need the cycle breaking check it doesn't suffer from having it.